plugins/thunderbolt/fu-plugin-thunderbolt.c

```diff
@@ -0,0 +1,736 @@
+		return TRUE;
+	}
+
+	if (g_str_equal (action, "add")) {
+		fu_plugin_thunderbolt_add (plugin, device);
+	} else if (g_str_equal (action, "remove")) {
+		fu_plugin_thunderbolt_remove (plugin, device);
+	} else if (g_str_equal (action, "change")) {
+		fu_plugin_thunderbolt_change (plugin, device);
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_plugin_thunderbolt_validate_firmware (GBytes *blob_fw, GError **error)
+{
+	/* FIXME: need to implement */
+	return TRUE;
+}
+
+static GFile *
+fu_plugin_thunderbolt_find_nvmem (GUdevDevice  *udevice,
+				  GError      **error)
+{
+	const gchar *devpath;
+	const gchar *name;
+	g_autoptr(GDir) d;
+
+	devpath = g_udev_device_get_sysfs_path (udevice);
+
+	d = g_dir_open (devpath, 0, error);
+	if (d == NULL)
+		return NULL;
+
+	while ((name = g_dir_read_name (d)) != NULL) {
+		if (g_str_has_prefix (name, "nvm_non_active")) {
+			g_autoptr(GFile) parent = g_file_new_for_path (devpath);
+			g_autoptr(GFile) nvm_dir = g_file_get_child (parent, name);
+			return g_file_get_child (nvm_dir, "nvmem");
+		}
+	}
+
+	g_set_error_literal (error,
+			     FWUPD_ERROR, FWUPD_ERROR_NOT_SUPPORTED,
+			     "Could not find non-volatile memory location");
+	return NULL;
+}
+
+static gboolean
+fu_plugin_thunderbolt_trigger_update (GUdevDevice  *udevice,
+				      GError      **error)
+{
+
+	const gchar *devpath;
+	ssize_t n;
+	int fd;
+	g_autofree gchar *auth_path = NULL;
+
+	devpath = g_udev_device_get_sysfs_path (udevice);
+	auth_path = g_build_filename (devpath, "nvm_authenticate", NULL);
+
+	fd = open (auth_path, O_WRONLY | O_CLOEXEC);
+	if (fd < 0) {
+		g_set_error (error, G_IO_ERROR,
+			     g_io_error_from_errno (errno),
+			     "could not open 'nvm_authenticate': %s",
+			     g_strerror (errno));
+		return FALSE;
+	}
+
+	do {
+		n = write (fd, "1", 1);
+		if (n < 1 && errno != EINTR) {
+			g_set_error (error, G_IO_ERROR,
+				     g_io_error_from_errno (errno),
+				     "could write to 'nvm_authenticate': %s",
+				     g_strerror (errno));
+			(void) close (fd);
+			return FALSE;
+		}
+	} while (n < 1);
+
+	(void) close (fd);
+	return TRUE;
+}
+
+static gboolean
+fu_plugin_thunderbolt_write_firmware (GUdevDevice  *udevice,
+				      GBytes       *blob_fw,
+				      GError      **error)
+{
+	gsize fw_size;
+	gsize nwritten;
+	gssize n;
+	g_autoptr(GFile) nvmem = NULL;
+	g_autoptr(GOutputStream) os = NULL;
+
+	nvmem = fu_plugin_thunderbolt_find_nvmem (udevice, error);
+	if (nvmem == NULL)
+		return FALSE;
+
+	os = (GOutputStream *) g_file_append_to (nvmem,
+						 G_FILE_CREATE_NONE,
+						 NULL,
+						 error);
+
+	if (os == NULL)
+		return FALSE;
+
+	nwritten = 0;
+	fw_size = g_bytes_get_size (blob_fw);
+
+	do {
+		g_autoptr(GBytes) fw_data = NULL;
+
+		fw_data = g_bytes_new_from_bytes (blob_fw,
+						  nwritten,
+						  fw_size - nwritten);
+
+		n = g_output_stream_write_bytes (os,
+						 fw_data,
+						 NULL,
+						 error);
+		if (n < 0) {
+			return FALSE;
+		}
+
```

Comment:
I think we want some progress notification here. Just keep a counter and then use `fu_plugin_set_percentage()` with a simple divisor.